### ISSUE TITLE
[Request] delay effect on microphone output.

### ISSUE THREAD

**Issue by altruios** [wwmm/easyeffects#753]
Can you please add the delay effect to output.
(And maybe fix up the delay effect variables).

And correct / increase delay length...such that it is displayed in either ms, or seconds (Currently i type in 1625 on the left or right channel and it jumps to 1.54... so not only do I think it's in seconds, there seems to be a rounding error of some kind happening... please use int for that millisecond variable -  not a floater, and I would love it to be in milliseconds, not seconds - the display says 'ms' but the numbers are getting converted to floats - so... clearly it's seconds)

My goal is self reflected harmony on discord. 2 extra voices that have decently long delays, 1625 and 3350 respectively... please - I have not found anything that is functionality on output. I have been hunting.

This is precisely the effect I am requesting. (queen: the prophet's song - 3:30 in)
https://youtu.be/HzdjMLKKdgk?t=210

**Comment by wwmm** [comment]
Putting this plugin in the microphone pipeline is easy. I can do that. But increasing the amount of delay is something that only its author can do as I do not have the necessary knowledge to do that. We use the delay from `Linux Studio Plugins` https://lsp-plug.in/?page=manuals&section=comp_delay_x2_stereo. You can contact its author in this page https://github.com/sadko4u/lsp-plugins. If he decides to make changes to the range I can update the PulseEffects interface with the new values.

The delay is in milliseconds. The maximum values supported by this plugin is `1000` ms. At this moment GTK default behavior is to go back to the previous valid value within the allowed range.

**Comment by altruios** [comment]
ah, copy - that would explain the snapback to 1.5... kinda but why does it snap back to a float? should it not be ints for that field? I'll ask the author that then... and ask them to increase the valid amount for a delay. aye... (also maybe adding TAPS for multiple individually editable delays...)

**Comment by wwmm** [comment]
>should it not be ints for that field

Although I could force int to be used in the interface at the end of the day the plugin is using floats.

**Comment by altruios** [comment]
his reply:
**

The compensation delay was developed for a bit another purposes: to adjust phase difference between two tracks. For example, you have drum multi-mic recording. To make drums sounding well, you need to adjust the delay for each track to make tracks in-phase. That's why the delay allows to set time offset in distance units (meters and centimeters), samples and milliseconds. Moreover, delay of 100 seconds requires huge memory buffers to be allocated (about 20 MB at 48 kHz sample rate).

Allocating such buffer for phase correction purposes is too generously.

I think better would be to implement another one delay plugin which will fit your requirements instead of modifying comp_delay_x2 plugin.

FYI, there's also another plugin which works with short delays and allows to make some spatial effects.
https://lsp-plug.in/?page=manuals&section=slap_delay_stereo**

so long story short - he is prob not going to implement any changes - but that delay effect may not even be appropriate for output. looking into a delay plugin for output was his suggestion.

**Comment by Digitalone1** [comment]
zamaudio has a delay plugin up to 8 seconds, but unfortunately it's only mono.

**Comment by altruios** [comment]
Only mono is better than nothing.

I haven’t looked at the code yet - but one crazy thought to make that bomb awesome:

A wrapper effect where: it’s an array of ‘tap’ that can be added/removed and each can be stereopanned. Inside the card is a new delay effect (or maybe any arbitrary effect?) with individual controls.

That may be pie in the sky

**Comment by wwmm** [comment]
Right now PulseEffects pipeline works only with stereo streams. Although it is possible to use a mono plugin things become way more complex. At some point in the past I had to do that. It is necessary to create create 2 instances of the mono plugin and deinterleaving the audio stream so we can process each channel individually in each plugin instance. After that we have to interleave the channels again so the stereo stream is restored. If possible I would like to avoid this solution as dealing with this process in a dynamic pipeline is painful.

**Comment by Digitalone1** [comment]
I don't know, but that should be something done at the plugin side. Too complex to be implemented in the PulseEffects pipeline.

In zamaudio repository issues there's also a request to make the dynamic equalizer available as stereo. At least the maximizer has two channels.

**Comment by altruios** [comment]
Aye, I didn’t know the specifics - maybe I can find a plug in like the one I’m thinking of already (stereo from mono)

**Comment by altruios** [comment]
changes have been made, and the long delay has been added :)
https://github.com/sadko4u/lsp-plugins/commit/819a54b2bef5e7453d424519a6fdac74dba13cb2

**Comment by Digitalone1** [comment]
Is it a new plugin or the new feature is just integrated in the existing delay?

**Comment by wwmm** [comment]
In EasyEffects 6.0.0 the delay plugin can be used in the mic pipeline.
